src/store/sled.rs

```diff
@@ -18,193 +15,196 @@ use libsignal_service::{
     push_service::DEFAULT_DEVICE_ID,
 };
 use log::{debug, trace, warn};
+use matrix_sdk_store_encryption::StoreCipher;
 use prost::Message;
-use sled::IVec;
+use serde::{de::DeserializeOwned, Serialize};
+use sled::Batch;
 
 use super::{ContactsStore, MessageStore, StateStore};
 use crate::{manager::Registered, proto::ContentProto, store::Thread, Error, Store};
 
-const SLED_KEY_REGISTRATION: &str = "registration";
 const SLED_KEY_CONTACTS: &str = "contacts";
+const SLED_KEY_PRE_KEYS_OFFSET_ID: &str = "pre_keys_offset_id";
+const SLED_KEY_NEXT_SIGNED_PRE_KEY_ID: &str = "next_signed_pre_key_id";
+const SLED_KEY_REGISTRATION: &str = "registration";
+const SLED_KEY_STORE_CIPHER: &str = "store_cipher";
 
+const SLED_TREE_DEFAULT: &str = "state";
+const SLED_TREE_PRE_KEYS: &str = "pre_keys";
+const SLED_TREE_SIGNED_PRE_KEYS: &str = "signed_pre_keys";
+const SLED_TREE_IDENTITIES: &str = "identities";
 const SLED_TREE_SESSIONS: &str = "sessions";
-const SLED_TREE_THREAD_PREFIX: &str = "thread";
+const SLED_TREE_THREAD_PREFIX: &str = "threads";
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct SledStore {
-    db: Arc<RwLock<sled::Db>>,
+    db: Arc<sled::Db>,
+    cipher: Option<Arc<StoreCipher>>,
 }
 
 impl SledStore {
-    pub fn new(path: impl Into<PathBuf>) -> Result<Self, Error> {
+    pub fn open(path: impl Into<PathBuf>) -> Result<Self, Error> {
+        Self::open_with_passphrase(path, None::<&str>)
+    }
+
+    pub fn open_with_passphrase(
+        path: impl Into<PathBuf>,
+        passphrase: Option<impl AsRef<str>>,
+    ) -> Result<Self, Error> {
+        let database = sled::open(path.into())?;
+        let cipher = passphrase
+            .map(|p| Self::get_or_create_store_cipher(&database, p.as_ref()))
+            .transpose()?;
         Ok(SledStore {
-            db: Arc::new(RwLock::new(sled::open(path.into())?)),
+            db: Arc::new(database),
+            cipher: cipher.map(Arc::new),
         })
     }
 
+    fn get_or_create_store_cipher(
+        database: &sled::Db,
+        passphrase: &str,
+    ) -> Result<StoreCipher, Error> {
+        let cipher = if let Some(key) = database.get(SLED_KEY_STORE_CIPHER)? {
+            StoreCipher::import(passphrase, &key)?
+        } else {
+            let cipher = StoreCipher::new()?;
+            #[cfg(not(test))]
+            let export = cipher.export(passphrase);
+            #[cfg(test)]
+            let export = cipher._insecure_export_fast_for_testing(passphrase);
+            database.insert(SLED_KEY_STORE_CIPHER, export?)?;
```

Comment:
Shouldn't the use store the cipher somewhere? What the point of storing it in sled?

Reply:
This is how it works, akin to how you'd store keys for full disk encryption. What you don't store is the passphrase.